### PR TITLE
http: allowing altering the x-envoy header prefix

### LINKED ISSUE #5363: Make X-Envoy configurable
Right now, all Envoy "internal" headers are prefaced with X-Envoy.  This works well for folks who are using Envoy as their own proxy, but less well for cloud where one might be "transparently" proxying customer traffic where the customer might be using Envoy as well, and clients might want their X-Envoy headers to traverse a cloud proxy hop.

It'd be nice if the preface could be configurable, so that folks could differentiate internal X-Our-Build-Of-Envoy headers from X-Customer-Envoy headers.

More likely specific to Google, we'd like different teams to be able to have their own prefaces as well.

We could do this a number of ways (shamelessly stolen from an internal email from @curiouserrandy)

* Compile time: Have a macro that defines the prefix (X-Envoy vs. X-Google vs. X-My-Company whatever) and create header char* constants with macro stringification/concatenation based on this macro.  The problem with this is that that macro needs to be defined when the files that define the full header names are compiled, and those files are deep inside of envoy.  So if we want a single different prefix for different internal builds, that means we'd need separate build trees for envoy for those different parts (so that the actual envoy compilation would be done with different flags).  We could probably do that with blaze magic, but it would be somewhat ugly.

* Run time.  Refactor envoy to create flags with the X-Envoy prefix early in envoy startup based on the value of a variable exported from envoy, and allow setting of that variable earlier in envoy startup.  

* Link time.  Like run time, have the headers created from some other variable.  That variable name is exported, and specified in a separate .cc file for different targets.  Thus various builds could have their won  foo_team_envoy_header_prefix_spec.cc with 'char* envoy_header_prefix_spec = "X-Foo-Team";' in it,The problem with this is that it goes against my understanding of good practice for blaze, which is to have declarations and definitions in the same target, so if someone tried to link envoy and forgot to specify this they'd get a link error and have to do it by hand.  Or maybe we could have just that file be a different build target for every different consumer of envoy?  There may be a way to make this work semi-cleanly.

Per @htuch in the same thread, we're leaning towards run-time - it'd add a tiny bit to start-up cost but wouldn't make inline header parsing more expensive, and it would give us the flexibility we need internally

### PR DESCRIPTION
**Pull request by alyssawilk**
Adding bootstrap configuration so Envoy can trust and generate x-random-prefix headers instead fo x-envoy headers

Risk Level: Medium
Testing: UT, sorta integration test
Docs Changes: n/a
Release Notes: yes
Fixes #5363

### CODE REVIEW

**Comment by alyssawilk** [comment]
FWIW I'm not a huge fan of this solution - while I think the release assert is sufficient to guard against the fragility it still feels uncomfortably fragile.  I'm not sure how else to do a runtime and maintain the header const singleton but I'm very open to suggestions even of invasive or toil-y workarounds.

File: source/common/http/headers.h [hunk 1]
```diff
@@ -5,15 +5,40 @@
...
+    return prefix_;
+  }
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    prefix_ = prefix;
+  }
+
+private:
+  const char* default_prefix_ = "x-envoy";
+  const char* prefix_ = default_prefix_;
```
Comment (mattklein123) [inline source/common/http/headers.h]: Is it possible to make this an absl::optional such that trying to read the prefix before it is set will crash? Essentially I'm wondering if we can set in all cases and then crash if we read before set. This way we are sure we don't have inconsistent values somewhere.

Another way to do this is to independently track when it is read, and then crash if the prefix is set after it has ever been read?

Comment (alyssawilk) [reply source/common/http/headers.h]: We can, but I don't think it gets us anything other than a billion test changes.  Right now we verify that in the one place we set it, that it gets set correctly, so if anyone used the header map beforehand and the values don't match we will crash at server start up.

If we go the route you suggest, we'll have to set the singleton in all the unit tests and integration tests which use headers and not server.cc which I think is more trouble than it's worth given we have crash-based validation either way.  WDYT?

Comment (mattklein123) [reply source/common/http/headers.h]: I'm just worried about this leaking outside of headers somehow and we don't notice an inconsistency which seems pretty bad. 

Would it require test changes if we did my second suggestion which is to see if the value has ever been read before it is set? I think that would require no changes to any tests, since it would only fail in the case where we read before set.

Comment (alyssawilk) [reply source/common/http/headers.h]: Oh sorry, I think that'll break less, possibly only the new integration test and if so I can work around.
looking

File: source/common/http/headers.h [hunk 2]
```diff
@@ -5,15 +5,48 @@
...
+    read_ = true;
+    return prefix_;
+  }
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    bool same_string = absl::string_view(prefix_) == absl::string_view(prefix);
+    RELEASE_ASSERT(!read_ || same_string,
+                   "Attempting to change the header prefix after it has been used!");
+    // Don't bother swapping if there is no actual change.
+    if (!same_string) {
```
Comment (mattklein123) [inline source/common/http/headers.h]: Won't this crash in the line above so this if statement will always be false?

Comment (alyssawilk) [reply source/common/http/headers.h]: no, because in the common case it hasn't been read yet.
that said, this is a hack to work around a test-only asan problem - let me handle it more cleanly for prod code.

File: test/integration/header_prefix_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+static const char* custom_prefix_ = "x-custom";
+
+class HeaderPrefixIntegrationTest : public HttpProtocolIntegrationTest {
+public:
+  static void SetUpTestSuite() {
+    ThreadSafeSingleton<Http::PrefixValue>::get().setPrefix(custom_prefix_);
+  }
+  void TearDown() {
+    // Because server.cc calls setPrefix using bootstrap, make sure to setPrefix with something
```
Comment (mattklein123) [inline test/integration/header_prefix_integration_test.cc]: Sorry I'm confused here. Why is this necessary? Would it be better to just store an `std::string` inside the prefix object instead of a reference to an external string?

Comment (alyssawilk) [reply test/integration/header_prefix_integration_test.cc]: I can, and that'd work fine for our use case, but from a thread safety perspective, saying "the underlying memory has to outlast the setPrefix call" makes sense where having a mutex "guarding" the underlying memory of the string which could be swapped out by another setPrefix call is really not clean.

Comment (mattklein123) [reply test/integration/header_prefix_integration_test.cc]: Sorry I don't follow? What's wrong with just copying the prefix internal to the object?

Comment (alyssawilk) [reply test/integration/header_prefix_integration_test.cc]: So I think there's a paradigm in C++ which is "this points to a static variable, which will be around" and the static data pointed to can change.

If we store the memory internally it's IMO weird, because the mutex no longer really works. I guess with the guards around "no writes after reads" it's safe but it feels really icky.

I'll make the icky-to-me change since the RELEASE_ASSERT technically makes my concern more style than anything else.

File: source/common/http/headers.h [hunk 4]
```diff
@@ -5,15 +5,48 @@
...
+    absl::WriterMutexLock lock(&m_);
+    read_ = true;
+    return prefix_.c_str();
+  }
+
+  // The char* prefix is used directly, so must be available for the interval where prefix() may be
+  // called.
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    RELEASE_ASSERT(!read_ || prefix_ == std::string(prefix),
```
Comment (mattklein123) [inline source/common/http/headers.h]: I'm sorry to be completely dense here, but why do we need the check to see if the prefix is being set to the same value? Can't we just check to see if it was ever read, and then just universally set on the next line?

Comment (mattklein123) [reply source/common/http/headers.h]: I guess this is just needed for the integration test, right? Maybe add a comment about that? It would be cool if there was a way to avoid this in product code just to allow the test, but if you can't think of anything easy I think a comment is fine. 

/wait

Comment (alyssawilk) [reply source/common/http/headers.h]: Yeah, either we need the teardown in the integration test, or we need to avoid double setting the string here.
I'd prefer we take this out and have the extra code in TearDown, but you previously objected to that.

Comment (mattklein123) [reply source/common/http/headers.h]: OK got it. Sorry for the runaround I think this is all my confusion. I'm fine with this with the comment, but if you prefer the other way, feel free to swap and just add an equally sufficient comment trail?
